shard: avoid calling shard function when possible

rkothiya opened the pull request:
When the pid is GF_CLIENT_PID_GSYNCD then just skip
calling any shard functions and wind the call.

fixes: #1702

Change-Id: Ia04e3586fb17159e1cb27348ac835af7771327a7
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

Commit:
shard: avoid calling shard function when possible

When the pid is GF_CLIENT_PID_GSYNCD then just skip
calling any shard functions and wind the call.

fixes: #1702

Change-Id: Ia04e3586fb17159e1cb27348ac835af7771327a7
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

mykaul commented on xlators/features/shard/src/shard.c:
```diff
@@ -1917,16 +1917,22 @@ shard_stat(call_frame_t *frame, xlator_t *this, loc_t *loc, dict_t *xdata)
-               uuid_utoa(loc->inode->gfid));
-        goto err;
-    }
+    if (frame->root->pid != GF_CLIENT_PID_GSYNCD) {
```
I'd do it differently:
if (frame->root->pid == GF_CLIENT_PID_GSYNC) {
    STACK_WIND(...)
    return;
}

...

Commit:
shard: avoid calling shard function when possible

Changed the sequence of checking the condition
as per the review comments.

fixes: #1702

Change-Id: Id909a4427d808fde4d1d58c9fe8097cad4f0b729
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

vh05 commented on xlators/features/shard/src/shard.c:
```diff
@@ -6939,6 +7005,35 @@ shard_common_inode_write_begin(call_frame_t *frame, xlator_t *this,
     uint64_t block_size = 0;
     shard_local_t *local = NULL;
 
+    if (frame->root->pid == GF_CLIENT_PID_GSYNCD) {
```
We also have this switch handling after getting the block size. To avoid code duplication, we can have an inline function for this

vh05 commented on xlators/features/shard/src/shard.c:
```diff
@@ -6452,15 +6486,26 @@ shard_common_remove_xattr(call_frame_t *frame, xlator_t *this,
+                            FIRST_CHILD(this)->fops->removexattr, loc, name,
+                            xdata);
+        else
+            STACK_WIND_TAIL(frame, FIRST_CHILD(this),
```
if (!loc), shouldn't we call fremovextttr just like we do in !block_size?

Commit:
shard: avoid calling shard function when possible

Added inline function for repeating code as per review comments
and called fremovexattr instead of removexattr in stack_wind_tail

fixes: #1702

Change-Id: I68c36cc3931acd9d8a0a3e2228f3fea0e34fdcfb
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/run regression

vh05 reviewed (approved):
Okay for me. Please wait for other reviewer

rkothiya commented:
@xhernandez  vinayak wanted one more review to be done for this patch, hence adding you as a reviewer.
